src/codec.rs

```diff
@@ -125,78 +154,126 @@ fn number_of_digits(mut n: usize) -> usize {
-            return Ok(None);
-        }
-
-        let (msg, len) = match parse_message(src) {
-            Ok((remaining, msg)) => (str::from_utf8(msg), src.len() - remaining.len()),
-            Err(Err::Incomplete(Needed::Size(min))) => {
-                self.remaining_msg_bytes = min.get();
-                return Ok(None);
-            }
-            Err(Err::Incomplete(_)) => {
-                return Ok(None);
-            }
-            Err(Err::Error(err)) | Err(Err::Failure(err)) => {
-                let code = err.code;
-                let parsed_bytes = src.len() - err.input.len();
-                src.advance(parsed_bytes);
-
-                match find_next_message(src) {
-                    Ok((_, position)) => src.advance(position),
-                    Err(_) => src.advance(src.len()),
+        let mut parse = Ok(());
+
+        // Parse the headers first if necessary
+        if self.headers_len.is_none() {
+            // Placeholder used for parsing headers into
+            let dst = &mut [httparse::EMPTY_HEADER; 2];
+
+            // Parse the headers and try to extract values
+            match httparse::parse_headers(src, dst) {
+                // A complete set of headers was parsed succesfully
+                Ok(httparse::Status::Complete((header_len, headers))) => {
+                    // If some headers were parsed successefully, set the headers length
+                    self.headers_len = Some(header_len);
+                    // Scan through the headers
+                    for header in headers {
+                        // If the "Content-Length" header is found, parse the value as a usize
+                        if header.name == "Content-Length" {
+                            match std::str::from_utf8(header.value) {
+                                // If there was an error parsing the "Content-Length" value as UTF-8, set the error and bail from the loop early
+                                Err(err) => {
+                                    parse = Err(ParseError::Utf8(err));
+                                    break;
+                                }
+                                // If there was an error parsing the "Content-Length" UTF-8 as a usize, set the error and bail from the loop early
+                                Ok(content_len) => match content_len.parse() {
+                                    Err(_) => {
+                                        parse = Err(ParseError::InvalidLength);
+                                        break;
+                                    }
+                                    // Otherwise, successfully set content_len from the parsed "Content-Length" value
+                                    Ok(content_len) => {
+                                        self.content_len = Some(content_len);
+                                    }
+                                },
+                            }
+                        }
+                    }
```

Comment:
I believe this `for`-`if` combo could be replaced with an `if let`-`find` instead? We only ever care about the `Content-Length` header, so there's no reason to keep scanning through the headers if there are more than one.

```suggestion
                    // If the "Content-Length" header exists, parse the value as a `usize`.
                    if let Some(header) = headers.iter().find(|h| h.name == "Content-Length") {
                        match std::str::from_utf8(header.value) {
                            Ok(content_len) => match content_len.parse() {
                                // Successfully set `content_len` from the parsed "Content-Length"
                                // value.
                                Ok(len) => self.content_len = Some(content_len),
                                // If there was an error parsing the "Content-Length" UTF-8 as a
                                // `usize`, set the error.
                                Err(_) => parse = Err(ParseError::InvalidLength),
                            }
                            // If there was an error parsing the "Content-Length" value as UTF-8,
                            // set the error.
                            Err(err) => parse = Err(ParseError::Utf8(err),
                        }
                    }
```